core/src/main/java/hudson/cli/CLIAction.java

```diff
@@ -91,46 +107,159 @@ public Object getTarget() {
-                            // capture the identity given by the transport, since this can be useful for SecurityRealm.createCliAuthenticator()
-                            channel.setProperty(CLICommand.TRANSPORT_AUTHENTICATION, Jenkins.getAuthentication());
-                            channel.setProperty(CliEntryPoint.class.getName(),new CliManagerImpl(channel));
+                        protected void onEncoding(String text) {
+                            try {
+                                encoding = Charset.forName(text);
+                            } catch (UnsupportedCharsetException x) {
+                                LOGGER.log(Level.WARNING, "unknown client charset {0}", text);
+                            }
+                        }
+                        @Override
+                        protected void onStart() {
+                            ready();
+                        }
+                        @Override
+                        protected void onStdin(byte[] chunk) throws IOException {
+                            stdinMatch.write(chunk);
+                        }
+                        @Override
+                        protected void onEndStdin() throws IOException {
+                            stdinMatch.close();
+                        }
+                        @Override
+                        protected void handleClose() {
+                            ready();
+                            Thread t = runningThread.get();
+                            if (t != null) {
+                                t.interrupt();
+                            }
+                        }
+                        private synchronized void ready() {
+                            ready = true;
+                            notifyAll();
+                        }
+                    }
+                    try (ServerSideImpl connection = new ServerSideImpl(upload, download)) {
+                        connection.begin();
+                        synchronized (connection) {
+                            while (!connection.ready) {
+                                connection.wait();
+                            }
+                        }
+                        PrintStream stdout = new PrintStream(connection.streamStdout(), false, connection.encoding.name());
+                        PrintStream stderr = new PrintStream(connection.streamStderr(), true, connection.encoding.name());
+                        if (connection.args.isEmpty()) {
+                            stderr.println("Connection closed before arguments received");
+                            connection.sendExit(2);
+                            return;
+                        }
+                        String commandName = connection.args.get(0);
+                        CLICommand command = CLICommand.clone(commandName);
+                        if (command == null) {
+                            stderr.println("No such command " + commandName);
```

Comment:
We could. But `CliManagerImpl` does not bother localizing that, so I am not going to bother now.